lib/mlir/Dialect/Clift/Utils/CBackend.cpp

```diff
@@ -0,0 +1,1247 @@
+      };
+    }
+
+    if (mlir::isa<BitwiseOrOp>(E)) {
+      return {
+        .Precedence = OperatorPrecedence::Bitor,
+        .Emit = &CEmitter::emitInfixExpression,
+      };
+    }
+
+    if (mlir::isa<LogicalAndOp>(E)) {
+      return {
+        .Precedence = OperatorPrecedence::And,
+        .Emit = &CEmitter::emitInfixExpression,
+      };
+    }
+
+    if (mlir::isa<LogicalOrOp>(E)) {
+      return {
+        .Precedence = OperatorPrecedence::Or,
+        .Emit = &CEmitter::emitInfixExpression,
+      };
+    }
+
+    if (mlir::isa<AssignOp>(E)) {
+      return {
+        .Precedence = OperatorPrecedence::Assignment,
+        .Emit = &CEmitter::emitInfixExpression,
+      };
+    }
+
+    if (mlir::isa<CommaOp>(E)) {
+      return {
+        .Precedence = OperatorPrecedence::Comma,
+        .Emit = &CEmitter::emitInfixExpression,
+      };
+    }
+
+    revng_abort("This operation is not supported.");
+  }
+
+  RecursiveCoroutine<void> emitExpression(mlir::Value V) {
+    const ExpressionEmitInfo Info = getExpressionEmitInfo(V);
+
+    bool PrintParentheses = Info.Precedence <= CurrentPrecedence
+                            and Info.Precedence != OperatorPrecedence::Primary;
+
+    if (PrintParentheses)
+      Out << '(';
+
+    // CurrentPrecedence is changed within this scope:
+    {
+      const auto PreviousPrecedence = CurrentPrecedence;
+      const auto PrecedenceGuard = llvm::make_scope_exit([&]() {
+        CurrentPrecedence = PreviousPrecedence;
+      });
+      CurrentPrecedence = Info.Precedence;
+
+      // Emit the expression using the member function returned by
+      // getExpressionEmitInfo.
+      rc_recur(this->*Info.Emit)(V);
+    }
+
+    if (PrintParentheses)
+      Out << ')';
+  }
+
+  RecursiveCoroutine<void> emitExpressionRegion(mlir::Region &R) {
+    revng_assert(R.hasOneBlock());
+    mlir::Block &B = R.front();
+
+    auto End = B.end();
+    revng_assert(End != B.begin());
+
+    auto Yield = mlir::cast<YieldOp>(*--End);
+    return emitExpression(Yield.getValue());
+  }
+
+  //===---------------------------- Statements ----------------------------===//
+
+  RecursiveCoroutine<void> emitLocalVariableDeclaration(LocalVariableOp S) {
+    // TODO: Emit variable name from the model once the model is extended to
+    //       provide this information.
+
+    rc_recur emitDeclaration(S.getResult().getType(),
+                             C.getVariableLocationDefinition(S.getSymName(),
+                                                             *CurrentFunction));
+
+    if (not S.getInitializer().empty()) {
+      Out << ' ' << '=' << ' ';
+
+      // Comma expressions in a variable initialiser must be parenthesized.
+      CurrentPrecedence = OperatorPrecedence::Comma;
+
+      rc_recur emitExpressionRegion(S.getInitializer());
+    }
+
+    Out << ';' << '\n';
+  }
+
+  RecursiveCoroutine<void> emitLabelStatement(AssignLabelOp S) {
+    Out.unindent();
+
+    Out << S.getLabelOp().getName() << ':';
```

Comment:
I added checks for uniqueness among the full set of local symbols introduced by local variables and labels. Note that this is not enough to preclude a conflict between a local symbol and a parameter.

Reply:
Ok that's a start, but I have some issues with the current implementation.
1. At the moment you detect a conflict if a local variable has the same name as a `goto`  label. That's not necessary. AFAIU block labels live in a separate namespace from any other symbol. See [this godbolt](https://godbolt.org/z/MY4dM3K4d). So if anything there should be 2 sets for checking uniqueness: one for label names, and one for everything else.
2. For `LocalVariableOp`, names are optional (they can be empty), and they are intended mainly for debug. I just wouldn't check them, because at the end of the day it's the C code generation logic (and namely `NameBuilder`) that will take care of making sure they don't collide with each other and with other stuff, e.g. function parameters, global objects.
3. In principle, the same principle I applied in the previous item to local variable names should apply to names of `goto` labels. But we haven't quite well defined how to treat them in the model and in the `NameBuilder`, so for now we make the names mandatory for `MakeLabelOp`, we check they are unique and non empty. When we properly handle them in the model and `NameBuilder` we'll fix this, lift the restriction, and leave the `NameBuilder` treat them properly.

Reply:
2. The local variable name is used for output by the C backend because there's no corresponding model object yet.

If we want to just generate names for all the locals (e.g. `var0`, `label0`, etc.) then I'd rather just remove the names from both locals and labels entirely.

Reply:
2. The local variable name is used for output by the C backend because there's no corresponding model object yet.

If we want to just generate names for all the locals (e.g. var0, label0, etc.) then I'd rather just remove the names from both locals and labels entirely.

Reply:
> If we want to just generate names for all the locals (e.g. var0, label0, etc.) then I'd rather just remove the names from both locals and labels entirely.

Agreed!

Reply:
And then of course we have to make sure that we always generate names for them that don't conflict with each other.

I'm double checking if we can already do something also about naming conflicts with parameters and global objects.

Reply:
Ok local variable names are still very poorly handled. Let's not get our ambitions too high in clift too for now.

Just copy the old backend for the names we generate for local variables. And kinda do the same with labels, just changing the prefix.

Reply:
I implemented this in a separate commit. `_var_0` and `_label_0` etc. I handled conflicts with parameters but not with globals.